20221012-add-type-restrictions-to-dsl-syntax.md

```diff
@@ -0,0 +1,171 @@
+```
+define viewer as self
+```
+we can now write this as
+```
+define viewer: [user]
+```
+
+The dropping of "self" will be extended to rewrite rules as well. If a rewrite contains `self` then it may be dropped as in the following examples:
+- `define viewer as self or viewer from parent` -> `define viewer: [user] or viewer from parent`
+- `define viewer as self and viewer from parent` -> `define viewer: [user] and viewer from parent`
+- `define viewer as self but not banned` -> `define viewer: [user] but not banned`
+
+### Wildcards
+
+The syntax and semantics of wildcards have also been updated in the new DSL. Consider the (new) authorization model
+```
+type user
+
+type document
+  relations
+    define viewer: [user]
+```
+Previously, writing a tuple with the wildcard symbol `document:X#viewer@*` meant that we intended the object `document:X` to be publically "viewable". In the new DSL, we will also add type restrictions to the wildcard symbol, and there will no longer be an untyped wildcard symbol. If you want to say that a document can be publically viewable for all `user`s, then you must add `user:*` to the type restrictions:
+```
+type user
+
+type document
+  relations
+    define viewer: [user, user:*]
+```
+Then you can write tuples such as `document:X#viewer@user:*`, which means that `document:X` is viewable to all `user`s. If you had another type, say `employee`, then you may instead write something like
+```
+type user
+
+type employee
+
+type document
+  relations
+    define viewer: [user, employee:*]
+```
+which means you could write tuples such as `document:Y#viewer@employee:*`, which means that `document:Y` is viewable to all `employee`s.
+
+## Example
+
+Consider the following authorization model:
+```
+type group
+  relations
+    define member as self
+
+type document
+  relations
+    define viewer as self
+```
+Suppose we want the following restrictions:
+1. viewer of documents can be users, groups, or members of a group
+1. some documents may be viewable for all users
+1. group members are users
+
+We can write the model with these type restriction as
+```
+model
+  schema 1.1
+
+type user
+
+type group
+  relations
+    define member: [user]
+
+type document
+  relations
+    define viewer: [user, group, group#member, user:*]
+```
+
+## Affect of type restrictions: writing tuples
+
+Given a model with type restrictions, we can now only write tuples that respect the type restriction. For example, given our model above the following are valid tuples:
+- object=group:eng, relation=member, user=user:alice
+- object=document:w, relation=viewer, user=user:beatrix
+- object=document:x, relation=viewer, user=group:eng
+- object=document:y, relation=viewer, user=group:hr#member
+- object=document:z, relation=viewer, user=user:*
+
+The following tuples would _not_ be valid:
+- object=group:eng, relation=member, user=charlie           # group members must have type user
```

Comment:
Minor nitpick on the words here.. Mind if we change to

group members must __be related to type `user`__

Reply:
Thanks!